Restore default empty skeleton path on MeshInstance3D

smix8 opened the pull request:
Restores default empty skeleton path on MeshInstance3D to avoid unnecessary or faulty NodePath updates.

_**I dont know if there are any additional quirks now that I am not considering. I asked in dev chat but this needs confirmation from the animation/editor people.**_

### How we got to this and why it is a problem

The original `MeshInstance` pre Godot 4 had an empty `NodePath` for `skeleton_path` as default.

This empty nodepath served 2 purposes. Because the NodePath was fully empty it did not try to resolve and get the node. It was also not prone to faulty and wasteful NodePath auto-updates on all reparent actions (an issue of its own) because there was nothing to update.

In Godot 4.0 that empty default was changed with the PR44630 to initalize class variables.

Instead of adding an empty `NodePath()` that PR changed to a relative`NodePath(..)` targeting the parent for reasons yet unknown.

Going further that change would turn the `MeshInstance3D` into a node that would try to resolve its parent node path for a skeleton all the time even if it was never used.

Another problem with the relative nodepath is that now this NodePath also gets auto-updated on reparent by NodePath and editor magic. If that fails you get error spam, lots and lots of error spam. Try finding that offending node with a broken path in a very large SceneTree when every second node you use is a MeshInstance3D with no skeleton seen within miles, good luck.

In the most ridiculous of all situations, if the relative nodepath would target scene `root`, this could try to resolve to `EditorRoot`, creating a new nodepath that spans from america to china causing a beauty of a commit diff because the path from EditorRoot just to the editor 3d viewport alone goes through 101 different Control related gui nodes.

### Compatibility 

~In order to keep compatibility with the current weirdness I made it look at the parent node when the `skeleton_path` is empty. If the parent node is a skeleton it is used as a fallback to simulate the old (..) parent nodepath behavior.~ If you relied on the default to parent path for your skeleton I have bad new for you. Godot does not store defaults, so with this PR that path would turn empty and you need to set the path manually again to fix it.

For existing projects that already have a convoluted path set in that property from a reparent nothing should change in behavior as that path is already stored for the node variable within the scene file. If they want to remove those unnecessary and often long and faulty nodepaths they need to clear them manually.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

smix8 committed:
Restore default empty skeleton path on MeshInstance3D

Restores default empty skeleton path on MeshInstance3D to avoid unnecessary or faulty NodePath updates.

TokageItLab commented:
Maybe related with https://github.com/godotengine/godot/issues/42267

patwork commented on scene/3d/mesh_instance_3d.cpp:
```diff
@@ -175,17 +175,21 @@ void MeshInstance3D::set_blend_shape_value(int p_blend_shape, float p_value) {
-				new_skin_reference = skeleton->register_skin(skin_internal);
-			}
+	Skeleton3D *skeleton = nullptr;
+	if (skeleton_path.is_empty()) {
+		skeleton = Object::cast_to<Skeleton3D>(get_parent());
+	} else {
+		skeleton = Object::cast_to<Skeleton3D>(get_node(skeleton_path));
+	}
```
I understand it's due to backward compatibility, but personally, I don't like it here. It looks like it won't be possible to create a MeshInstance3D as a child of a skeleton without being "automagically" forced to use it (other than by selecting a different skeleton).

aaronfranke replied:
@patwork Can you explain your use case? If a mesh without any bones "uses" a Skeleton3D, it won't do anything.

I'd argue this change is a good idea even without compatibility concerns. Using the parent as a skeleton by default makes sense to me, which means using the parent when the NodePath is empty.

smix8 replied:
I was in the middle of updating the description lyuma.

patwork replied:
OK, my only concern is that the engine would force-connect the parent skeleton to the mesh without asking the user. In the editor, the skeleton node would be 'null', but this would change behind the scenes after running the project. This could be confusing for the user.

I guess there are two options:

1. leave force-connect (backward compatible) but state in documentation that engine will do it automatically for the user
2. remove backward compatibility, but warn users about it in migration guide (and also check other places in the engine if they are setting skeleton path correctly, like scene importer)

patwork replied:
because "force" ;) And because it's done in the runtime and not clearly visible in the editor. If the runtime connects mesh instances when their skeleton paths are null, it will not be possible to disable skeleton connection when the user wants it (other than moving specified mesh somewhere else in the tree).

smix8 replied:
Having auto-to-parent attachment behavior can work with node classes that are very narrow in scope, e.g. NavigationAgent attaching to a parent or collision shapes attaching to the body node.

For a node class as broadly used like MeshInstance3D for many things it is just bad design and a tripwire. MeshInstance3D is more or less the base not for anything visual that people use in projects, addons, and what not at very inflated amount.

smix8 commented:
Removed the contested compatibility code (I agree, it is not a good solution bot the other solution is accepting the break) and updated the PR description.

TokageItLab commented:
I propose the following in order to meet both needs while preserving compatibility:

First, add `Use External Skeleton` property. Default is `false`. It should work like old behavior, but path is not exposed in the inspector and if there is no parent skeleton, it should notice configure warning.

Then, you can implement migrating code `_set()` to keep compatibility. It means if skeleton path is not `..`, you can set `Use External Skeleton` to `false` in `#ifdef DISABLE_DEPRECATED` block.

smix8 commented:
> First, add Use External Skeleton property. Default is  .....

I appreciate all the thought put into this to somehow-someway preserve compatibility.

Imo this adds way too much convolutedness to the `MeshInstance3D` for something that should be a clear cut behavior requiring no extra documentation and explaination and behind the scene magic behavior.

You want to use a skeleton on a meshinstance? You set a path to that skeleton deliberately. End of (documentation) story.

This new compatibility stuff would become mostly irrelevant soon after release but would stay and persist to annoy and confuse users which is not a good tradeoff.

Imo this should be done as a clear cut compatibility break to fix the review hickup from Godot 4.0, else you are dragging this technical messup for all eternity for other generations to enjoy. 

If you ask me the `MeshInstance3D`, a most generic node used at plenty in 3d projects, shouldnt even be responsible for anything from a comparably niche-node like `Skeleton3D` in the first place. That is completely flipped responsibility and a design  issue but also a topic for another day.

TokageItLab reviewed (changes requested):
I agree with setting the default to empty somewhat. However, since compatibility issues are a concern, so I think at least the following two things are necessary if we make default to emply as break compat in some extent:

- We need to add a test to verify that the Skeleton Path is assigned correctly during import
- To maintain compatibility with past projects, we will provide a custom node such as SkinMeshInstance3D that automatically searches for the parent skeleton while preserving past behavior. (I guess this could be an add-on or example code rather than being added to the core and can be separate PR from this)

---

However, if there is no SkeletonPath to begin with, properties such as Skin may not be necessary. Considering that, I think the correct approach would be to divide the classes as mentioned [above](https://github.com/godotengine/godot/pull/107526#discussion_r2147339511) into MeshInstance3D and SkinMeshInstance3D, and have SkinMeshInstance3D perform the old MeshInstance3D search behavior (then if SkinMeshInstance3D is not a Skeleton child, a warning should be displayed on SceneTreeDock and can implement use External Skeleton option as I commented in https://github.com/godotengine/godot/pull/107526#issuecomment-2973841172). Then MeshInstance3D doesn't need to have `skeleton_path` and `skin` properties to begin with.

Eventually, the problem is that the current MeshInstance3D enforces the intended behavior while being multifunctional only for Skin, even though there are several use cases. Therefore, these classes should be divided according to their roles, and their behavior should be made optional.

So I believe the dividing class approach is the most appropriate way. However, a migration manual will be necessary in any case. If the timing is 5.0, I expect that it will be possible to convert classes depending on whether or not `skeleton_path` is used by using the project converter, but if this is done during 4.x, it will be necessary to rely on the manual. (but as long as the scene is imported from an external source, I believe it can be corrected by reimporting)

---

Based on above, in 4.x, I recommend to implement the three options as enum SkinUsage to MeshInstance3D:

- `SkinUsage`
  - `SKIN_USAGE_NONE`
  - `SKIN_USAGE_USE_PARENT_SKELETON`
  - `SKIN_USAGE_USE_EXTERNAL_SKELETON`

You can make `SKIN_USAGE_NONE` default, at this point, if the .tscn have skeleton_path, you should be able to set `SKIN_USAGE_USE_EXTERNAL_SKELETON` in the `#ifndef DISABLE_DEPRECATED` block in the setter. And at last, provide a manual that sets `SKIN_USAGE_USE_PARENT_SKELETON` only for those who have projects that depend on search operations.

and, at the time of the 5.0 launch, replace the `SKIN_USAGE_NONE` option with a MeshInstance3D that does not have a skin, and split the `SKIN_USAGE_USE_PARENT_SKELETON` and `SKIN_USAGE_USE_EXTERNAL_SKELETON` options into an extended class called SkinMeshInstance3D. I believe this is a reasonable compromise for minimizing compatibility breaks.

smix8 commented:
Closing in favor of https://github.com/godotengine/godot/pull/112267. That PR adds a compatibility project setting that needs to be enabled deliberate by users. Kinda agreed on that to avoid having intrusive or "magic" compatibility behavior going further.
